### PR TITLE
Fuzzing tests for ledger rpc

### LINKED ISSUE #397: Add proptest for ledger RPC
### Description
The PR #394 added a test structure for ledger rpc but does not support fuzzy testing yet. This issue proposes to add fuzzing to ledger rpc to make the testing suite more robust.

The fuzzing should in particular test the different query modes ("Standard", "Compact" and "Full"), the different identifier types (by number, hash or id) and parameter flattening (see README for ledger_getBatches for instance)

### PR DESCRIPTION
**Pull request by theochap**
# Description
This PR implements fuzzing tests for ledger rpc. The fuzzing api is used to generate random test data for the ledger db and random queries to test the ledger rpc api.

## Linked Issues
- Fixes #397

## Testing
Some tests have been added to the test_rpc.rs folder.

### CODE REVIEW

**Review by bkolad — APPROVED** [review]

**Comment by theochap** [comment]
Followed a bug in the CI I opened the issue #417 to report it. I removed the associated test from the suite

File: examples/demo-rollup/Cargo.toml [hunk 1]
```diff
@@ -44,3 +44,5 @@ sha2 = { workspace = true }
 reqwest = "0.11"
 tendermint = "0.32"
 tempfile = { workspace = true }
+proptest = { workspace = true}
```
Comment (citizen-stig) [inline examples/demo-rollup/Cargo.toml:47]: Formatting
```suggestion
proptest = { workspace = true }
```

File: examples/demo-rollup/src/test_rpc.rs [hunk 2]
```diff
@@ -11,69 +14,38 @@ use tokio::sync::oneshot;
 
 use crate::{config::RpcConfig, ledger_rpc};
 
-async fn query_test_helper(data: String, expected: &str, rpc_config: RpcConfig) {
+struct TestExpect {
+    data: String,
```
Comment (citizen-stig) [inline examples/demo-rollup/src/test_rpc.rs:18]: Nit, a suggestion, `data` might sound too generic, maybe "payload"?

File: examples/demo-rollup/src/test_rpc.rs [hunk 3]
```diff
@@ -83,44 +55,102 @@ fn test_helper(data: String, expected: &str) {
...
 
-        query_test_helper(data, expected, rpc_config).await;
+        query_test_helper(test_queries, rpc_config).await;
+
+        // By closing the `TempDir` explicitly we can check that it has
+        // been deleted successfully. If we don't close it explicitly,
+        // the directory will still be deleted when `tmp_dir` goes out
+        // of scope, but we won't know whether deleting the directory
+        // succeeded.
+        tmpdir.close().unwrap();
```
Comment (citizen-stig) [inline examples/demo-rollup/src/test_rpc.rs:93]: But isn't that the whole purpose of the `tempdir` package? 

Also, even if deleting directory fails, should it fail our test?

File: examples/demo-rollup/src/test_rpc.rs [hunk 4]
```diff
@@ -152,30 +207,347 @@ fn test_get_batches() {
...
+                    }
+                    curr_tx_num += 1;
+                }
+
+            }
+
+        }
+
+        let data = format!(r#"{{"jsonrpc":"2.0","method":"ledger_getEvents","params":[{random_event_num}],"id":1}}"#);
+        let expected : String= r#"{"jsonrpc":"2.0","result":[null],"id":1}"#.to_string();
```
Comment (citizen-stig) [inline examples/demo-rollup/src/test_rpc.rs:548]:
```suggestion
        let expected: String = r#"{"jsonrpc":"2.0","result":[null],"id":1}"#.to_string();
```

Comment (citizen-stig) [reply examples/demo-rollup/src/test_rpc.rs:548]: It looks like formatting inside macro isn't working...

File: rollup-interface/src/state_machine/crypto/simple_merkle.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
+pub const DEFAULT_LEAF_DOMAIN_SEPARATOR: [u8; 1] = [0];
+pub const DEFAULT_INTERNAL_DOMAIN_SEPARATOR: [u8; 1] = [1];
+
+pub trait MerkleHasher {
```
Comment (citizen-stig) [inline rollup-interface/src/state_machine/crypto/simple_merkle.rs:4]: Could you please add short description, why this trait is needed?

File: examples/demo-rollup/src/test_rpc.rs (hunk 4 above)
Comment (neysofu) [reply examples/demo-rollup/src/test_rpc.rs:548]: There's no way around this, sadly. We could use attribute macros (https://docs.rs/proptest-attr-macro/latest/proptest_attr_macro/ or https://github.com/proptest-rs/proptest/issues/153 once it's released) but it's not worth it IMO.

File: examples/demo-rollup/src/test_rpc.rs (hunk 3 above)
Comment (citizen-stig) [reply examples/demo-rollup/src/test_rpc.rs:93]: Let's remove this

File: rollup-interface/src/state_machine/crypto/simple_merkle.rs (hunk 5 above)
Comment (citizen-stig) [reply rollup-interface/src/state_machine/crypto/simple_merkle.rs:4]: @preston-evans98

File: .github/workflows/rust.yml [hunk 6]
```diff
@@ -88,7 +88,7 @@ jobs:
         with:
           version: "v0.4.0"
       - name: Run cargo test
-        run: cargo test
+        run: cargo test || find -type f -iname '*.proptest-regressions' -exec cat {} +
```
Comment (citizen-stig) [inline .github/workflows/rust.yml:91]: I believe we don't need this anymore:

```suggestion
        run: cargo test
```

**Review by neysofu — APPROVED** [review]

File: .github/workflows/rust.yml (hunk 6 above)
Comment (neysofu) [reply .github/workflows/rust.yml:91]: I believe that's correct, I'll commit your suggestion.

**Review by citizen-stig — APPROVED** [review]

File: rollup-interface/src/state_machine/crypto/simple_merkle.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+			fn inner_hash(&mut self, left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
+				let mut hasher = <$name as SimpleHasher>::new();
+				hasher.update(& $crate::crypto::simple_merkle::DEFAULT_INTERNAL_DOMAIN_SEPARATOR);
+				hasher.update(left);
+				hasher.update(right);
+				hasher.finalize()
+			}
+		}
+	}
+);
```
[the remark this answers is not in the document]
Comment (neysofu) [reply rollup-interface/src/state_machine/crypto/simple_merkle.rs]: Removed in https://github.com/Sovereign-Labs/sovereign-sdk/pull/403/commits/d2cd0bf0879734812e247cc5f5e05a1e2a759739; marking as resolved.

**Review by neysofu — APPROVED** [review]

File: rollup-interface/src/state_machine/crypto/simple_merkle.rs (hunk 5 above)
Comment (neysofu) [reply rollup-interface/src/state_machine/crypto/simple_merkle.rs]: I haven't yet added a doc comment, but in the meantime I moved it into `fuzzing.rs` so it's less present in the public API.

**Pull request merged**